src/cargo/util/toml/targets.rs

```diff
@@ -294,18 +294,17 @@ fn to_bin_targets(
 
     // This loop performs basic checks on each of the TomlTarget in `bins`.
     for bin in &bins {
+        validate_bin_name(bin, warnings)?;
+
         // For each binary, check if the `filename` parameter is populated. If it is,
         // check if the corresponding cargo feature has been activated.
         if bin.filename.is_some() {
             features.require(Feature::different_binary_name())?;
         }
 
-        validate_target_name(bin, "binary", "bin", warnings)?;
-
-        let name = name_or_panic(bin).to_owned();
-
         if let Some(crate_types) = bin.crate_types() {
             if !crate_types.is_empty() {
+                let name = name_or_panic(bin).to_owned();
```

Comment:
I reduced how much this was being reused and in weighing that against how specialied the look up was, I felt it'd be nicer to keep it closer to where its used, rather than having one variable used the whole function.